### ISSUE TITLE
Running Simple Publisher and subscriber

### ISSUE THREAD

**Issue by Viplav04** [micro-ROS/micro-ros-build#18]
@iluetkeb
I am trying the built in publisher and subscriber. I have only connected the board with USB OTG and I use ACM profile.
The publisher node is running in the nsh. But there is no printing publisher or anything in the nsh. Or will it show on the agent?

**Comment by Viplav04** [comment]
I changed the config file and set my serial device as /dev/ttyACM0. The nuttshell works. But running the agent is not working after typing ./micro_ros_agent serial /dev/ttyACM0. It connects to /dev/ttyUSB0 which is my JTAG Debugger. 
I also tried udp way, setting the agent udp port 8888. It initializes. Also the IP 127.0.0.1 gives a ping. But telnet fails on it. Maybe I am doing something wrong with the UDP configuration.

Can you give some instructions to connect the client and agent by USB OTG or UDP and running the publisher. When I run the publisher, the green light blinks on the board. I am assuming the publisher is working.

**Comment by iluetkeb** [comment]
> I changed the config file and set my serial device as /dev/ttyACM0.

rmw_microxrcedds.config is the configuration for the client, i.e. in the firmware. Here, you have to use NuttX device names, which are /dev/ttyS0 for the first serial, /dev/ttyS1 for the second, and so on.

> typing ./micro_ros_agent serial /dev/ttyACM0. It connects to /dev/ttyUSB0 

That is very odd. There is no default filename for the serial device (I've checked the source code). What you ran should have worked. How have you found out that it connects to ttyUSB0?

Regarding IP-based communication:
 - 127.0.0.1 is, by convention, the IP address of your own machine. so, yes, you can ping it, but it's your own computer, not the embedded board.
 - NuttX can use both DHCP and static IP configuration. If you have connected the board to a network which has a DHCP server, you can enable that in the NuttX configuration. You find that in `Application Configuration/NSH Library/Networking Configuration/Network initialization`
 - If you don't want or can't use DHCP, you can statically configure an IP address for the board and the gateway (-> either your router, or if directly connected, your own machine). It's in the same place, just below the DHCP check box. Beware that you have to convert the IP addresses to hex.

**Comment by Viplav04** [comment]
My Nuttx uses serial /dev/ttyACM0 , that is what I defined in the config file. But when I run my agent it doesn't print serial initialisation...OK but connecting the agent to ttyUSB0 which is the JTAG does print serial initialisation...OK.

**Comment by iluetkeb** [comment]
> My Nuttx uses serial /dev/ttyACM0 , that is what I defined in the config file.

Just to be sure what we're talking about. The following setup is what you want to achieve, right?
```
/----------------------\                  /---------------------------\
|      STM32-E407      |                  |      Linux                |
|----------------------|                  |---------------------------|
| JTAG   UART3      ---(    JTAG-Probe    ) --- /dev/ttyUSB0  OpenOCD |
| publisher UART6   ---( TTL232-USB Cable ) --- /dev/ttACM0   agent   |
-----------------------|                  |---------------------------|
```

I've used `UART3` and `UART6`, because that's what our default configuration uses. *If* you have only those two devices, the internal NuttX name (what you have to put into the config) for UART3 is `/dev/ttyS0` and for UART6 it is `/dev/ttyS1`

My first question would be: What's your nsh running on? Because if you have only those two devices, you can't run a nsh in parallel to the rest.

>  But when I run my agent it doesn't print serial initialisation...OK

Maybe it's a permission problem. Do a `ls -l /dev/ttyACM0` and check that you have write permissions. Maybe you need to

**Comment by Viplav04** [comment]
Yes, the setup is right but I did not connect UART3 and UART6. My nsh can now run both on serial /dev/ttyACM0 and telnet 10.0.0.2. I reconfigured the board. 
I wanted to publish from the board and subscribe it on my Linux as a first case.
I have read and write permission for ttyACM0.

Now both UDP and serial show initialisation OK.

Linux
Terminal 1
>telnet 10.0.0.2
Trying 10.0.0.2...
Connected to 10.0.0.2.
Escape character is '^]'.

NuttShell (NSH)
nsh> 

Linux Terminal 2
./micro_ros_agent serial /dev/ttyACM0
Serial agent initialization... OK
Enter 'q' for exit

**Comment by iluetkeb** [comment]
So, that looks good. What's missing?

**Comment by Viplav04** [comment]
Running the built in publisher example just keeps blinking a light on my board . Seeing the code of the publisher, I am expecting it to print something.

**Comment by iluetkeb** [comment]
Maybe you know this already, but the name of the device on the Linux side is different than the name of the device on the NuttX side. You cannot use /dev/ttyACM0 on both sides. On the NuttX side, you have to chose a name which corresponds to the UART you're using.

Alternatively, seeing that you new have working IP connectivity, you can also use udp and put the Linux machine's IP address in the config.

**Comment by Viplav04** [comment]
I did put the Linux's id 10.0.0.1 in the config. My firmware id is 10.0.0.2. 
The rmw config file should have 10.0.0.1 right?

**Comment by iluetkeb** [comment]
Yes, the config file should have 10.0.0.1, if that's the Linux's IP address.

btw, once the connection works, you will also see an output on the agent that a topic is being published.

**Comment by Viplav04** [comment]
Terminal 1:
telnet 10.0.0.2
Trying 10.0.0.2...
Connected to 10.0.0.2.
Escape character is '^]'.

NuttShell (NSH)
nsh> publisher

No output, blinking light at board

Terminal 2:
./micro_ros_agent udp 8888
UDP agent initialization... OK
Enter 'q' for exit
No output

Anything to be done at the agent side as well or?

Rmw config

CONFIG_MICRO_XRCEDDS_TRANSPORT=udp
CONFIG_IP=10.0.0.1
CONFIG_PORT=8888

CONFIG_DEVICE="/dev/ttyACM0"


CONFIG_MICRO_XRCEDDS_CREATION_MODE=xml

CONFIG_MAX_HISTORY=4
CONFIG_MAX_NODES=2
CONFIG_MAX_PUBLISHERS_X_NODE=4
CONFIG_MAX_SUBSCRIPTIONS_X_NODE=4
CONFIG_RMW_NODE_NAME_MAX_NAME_LENGTH=128
CONFIG_RMW_TOPIC_NAME_MAX_NAME_LENGTH=50
CONFIG_RMW_TYPE_NAME_MAX_NAME_LENGTH=128

**Comment by iluetkeb** [comment]
That all looks correct (except for the blinking light, I haven't seen that before, and, of course, the device name, but that isn't needed for udp).

Have you made sure everything was recompiled and flashed correctly?

**Comment by Viplav04** [comment]
Yes, as you said I deleted the rmw build and install directories.
Should I clean all build and install from both agent and firmware and try to build again?

**Comment by iluetkeb** [comment]
You don't have to do anything on the agent side, but for the firmware, you might want to go to "firmware/NuttX" and do a "make distclean". After that you can compile normally again.

**Comment by Viplav04** [comment]
Still the same after trying all of it again. The light is blinking at the board. Maybe is there something wrong with the publisher example file? Is there any other way to check the the client agent connection?

**Comment by Viplav04** [comment]
I realised I was defining a new IP for my Linux so I changed that. And the agent was connecting to the internet by UDP .
Now I turned off the internet and redefined the configurations as below:
My Linux machine IP address by default is 192.168.101.100.

Linux terminal 1:
telnet 192.168.101.101
Trying 192.168.101.101...
Connected to 192.168.101.101.
Escape character is '^]'.

NuttShell (NSH)
nsh> 

Linux terminal 2:
./micro_ros_agent udp 8888
UDP agent initialization... ERROR


CONFIG_MICRO_XRCEDDS_TRANSPORT=udp
CONFIG_IP=192.168.101.100
CONFIG_PORT=8888

CONFIG_DEVICE="/dev/ttyACM1"

UDP agent initialisation becomes OK if I turn on my internet connection through another wire. But then it is connecting to the wrong connection not to the board.

**Comment by Viplav04** [comment]
My agent is connecting to the wrong MAC address I guess. Is there a way to tell the agent to use a different MAC address?

**Comment by Viplav04** [comment]
I was doing something wrong. I will close this issue now. I didn't connect another connection for the publisher to send data. There were only 2 connections:
1 ethernet : running the agent via UDP
1 miniUSB - ACM profile running the client works
but the other way around does not work.

I will add another connection for the publisher output, I don't have the UART type connections so I will try to connect the linux with another miniUSB or attach an external display to the board to see the messages.
If you have any other ideas , please do share. I am new to this so I will probably do a lot of mistakes going ahead as I did now with the client agent connection.

**Comment by Viplav04** [comment]
I can't make the UART connections because of hardware constraints.

I am trying 3 connections
1. USB FS agent - /dev/ttyACM0
2. UDP client - telnet to run nuttshell
3. USB HS publisher - configure another USB HS connection to publish the messages.

Am I going the right way? Or is also possible to run both USBs as agent and client respectively and publish via UDP?

**Comment by jfm92** [comment]
Hi Viplav04,
I have a few questions about your set-up:

- How do you want to see the NSH console? 
There are several ways, to use the console. Connecting an USB cable to the USB-OTG1 port, by telnet or by the UART2/3. By default to make it easier, is configured to be used on the USB-OTG1. You can see a tutorial here: https://micro-ros.github.io/docs/tutorials/advanced/nuttx_getting_started/#NSH-USB-Olimex

- Which communication protocol do you want to use in micro-ROS? 
micro-ROS allows you to do UDP, TCP and Serial connections between the Client and the Agent. But is very important to know, Client and Agent MUST use the same protocol. I mean, if you want to use a micro-ROS client using serial communication, the agent must be configured to use serial and connect to the specific port.
In the next tutorial you can see how to do the right configuration:
https://micro-ros.github.io/docs/tutorials/advanced/nuttx_getting_started/#micro-ROS-demo-Olimex

Also is important to know, that at this moment by default is not available the second USB-OTG of the board and only the first one is available for the NSH console.

I would highly recommend you to see the second tutorial, but is probably that the Docker doesn't work due to the recent changes on the build system. But we're working on it.

**Comment by Viplav04** [comment]
Hi @jfm92 
Thank you for replying.
I don't have the UART hardware.
So I am thinking of using NSH on my serial console and UDP connection between client and agent.
I want to run the publisher example and subscribe it on my linux.
The issue with the UDP connection was that the Agent is connecting to my internet access MAC and not the USB Ethernet MAC which I have created locally to connect the board.
Is there a way to change that MAC address because when I remove my wired internet connection, the UDP agent initialisation gives an ERROR.
After resolving this step, I need to think how to publish the data as the second USB-OTG is not available as you said. So I have to find a way to send the publisher data to my linux.

**Comment by jfm92** [comment]
HI @Viplav04 ,

I'm sorry, but I don't fully understand your set-up.
At first, you're trying to access to the console through telnet. Then you have a micro-ROS client on the board and you want to use it through UDP.
On the computer you are running a micro-ROS Agent which is also configured to use UDP, right?

The blinking green led means that the memory is overflowing. Probably you have some problem on your configuration.

Anyways, I would highly recommend you to follow [this basic tutorial](https://micro-ros.github.io/docs/tutorials/advanced/nuttx_getting_started/#micro-ROS-demo-Olimex) to get familiarize with micro-ROS.

**Comment by Viplav04** [comment]
Okay, I have found a pl2303 driver to connect to USART3 and connected the terminals as mentioned in the tutorials. 
Now I have following connections:
1. USB - OTG - /dev/ttyACM0
2. PL2303 USB OTG - /dev/ttyUSB0
3. UDP with board connected by Ethernet and with Ethernet - USB converter at Linux. In Nuttx config
board IP - 10.0.0.2
Router IP - 10.0.0.1

RMW config file has 

CONFIG_MICRO_XRCEDDS_TRANSPORT=udp
CONFIG_IP=10.0.0.1
CONFIG_PORT=8888

CONFIG_DEVICE="/dev/ttyUSB0" 

Now, I can run the nsh by USB-OTG /ACM0. and connect the agent at ttyUSB0 right?

**Comment by Viplav04** [comment]
Okay I made the serial run. I was doing something wrong with the configuration.
The serial initialization says OK, but running the publisher still gives a blinking light and no printing.

My Nsh says:
nsh>free
             total       used       free    largest
Umem:       157632      25120     132512      89904

What things should I check in the configuration as I used the driverbase config. and just changed the USB driver settings.
Now my Nsh runs on /dev/ttyACM1, my UART has ttyS0 and the agent is connected to it in linux by ttyUSB1.

**Comment by Viplav04** [comment]
I got it to work!
Thank you for your time and comments. It would be better if you could update the build and other tutorials. They lack all steps for beginners, but after learning about Nuttx configurations, we can play around.

**Comment by iluetkeb** [comment]
>  It would be better if you could update the build and other tutorials. They lack all steps for beginners

Do you have concrete suggestions on what we are missing? We will work on them of course, but any concrete issues would help us.

**Comment by Viplav04** [comment]
I am new to the software world, I am studying mechatronics and my Bachelor's was in Mechanical engineering. So all of ROS and embedded things are new for me. Maybe that is why it took a lot of time for me. Anyway these are the issues where I think things can become easier:
During Build : I think we covered most of the build issues in the other thread. The instructions to build are clear, maybe some dependencies need to be defined first as mentioned in that thread.
Running the tutorials for micro-Ros and Nuttx config:
There seems to be steps missing or they are in the later parts as you go down in this page : https://micro-ros.github.io/docs/tutorials/advanced/nuttx_getting_started/#NSH-USB-Olimex
Also, some more information would be nice without the docker build, to explain how the connections work and how to change the configurations. Also you can refer to the Nuttx tutorials link on youtube : https://www.youtube.com/channel/UC0QciIlcUnjJkL5yJJBmluw/videos.

**Comment by Viplav04** [comment]
Can you also give some information to measure the real time performance of the publisher and subscriber example and how to change the different Quality of service settings there and measuring their performance.

**Comment by KevinHwangAi** [comment]
Hi @Viplav04. I had the same problem as you.

 When I entered the command "publisher" in nsh, the green light of  blinks on the stm32e407,and nothing appeared. 

Could you tell me how do you solve this problem?

**Comment by Viplav04** [comment]
Hi @KevinHwangAi 
It is because of something wrong in the configuration. I could not exactly pin point to what is causing it, but i played around with the configuration and got it done.
Best way is to configure using scripts/configure.sh olimex... nsh
Then manually enable
clock monotonic in RTOS Features,
Networking support if you are using through UDP, I use serial mode through UART3,
Enable micro ROS in applications and set directory as mcu_ws, then enable examples publisher and subscriber.
Also micro ROS requires uClibxx enabled in NSH library, i guess.
Then build using ros2 run micro_ros_setup build_firmware.sh

**Comment by iluetkeb** [comment]
@Viplav04 that's an interesting report. could you "make savedefconfig" in your NuttX directory, and post the resulting "defconfig" file here? I would like to compare what is different with respect to our default configuration.

**Comment by Viplav04** [comment]
[defconfig.txt](https://github.com/micro-ROS/micro-ros-build/files/3706343/defconfig.txt)

Does it work or should I copy paste the contents of the file?

**Comment by Viplav04** [comment]
@iluetkeb Can you also tell if I am missing something in the configuration which can cause poor real time performance?

**Comment by KevinHwangAi** [comment]
@Viplav04 Thank you for your help. I used your defconfig file and brushed it into OLIMEX MCU. In NSH, after enter "publisher" and the green light still blinking . I guess the reason may be that I didn't use the correct PL2303 driver?

**Comment by Viplav04** [comment]
@KevinHwangAi did you set the serial device as /dev/ttyS0 in microxrcedds config file? then you have to check the connections as mentioned in the tutorial : 
USART3 TX -> PL2303 RX
USART3 RX -> PL2303 TX
GND Board -> Pl2303 GND

> The publisher will produce output on stdout as well when it's running.
> 
> However, the first step is to contact the agent. Before that step succeeds, nothing will be printed.
> 
> Please note that the configuration how to reach the agent is located in `firmware/mcu_ws/uros/rmw_microxrcedds/rmw_microxrcedds_c/rmw_microxrcedds.config`. You have to edit it manually and then recompile (and maybe remove the install and builds for rmw_microxrcedds_c before, I'm not sure the dependencies are correct, yet).
> 
> Yes, that's not ideal, and we'll expose it as a NuttX config option, and make it dynamic later, but right now, editing that file is what you need to do...

**Comment by KevinHwangAi** [comment]
@Viplav04 I set it in the file to /dev/ttyUSB0.Is it wrong?

**Comment by iluetkeb** [comment]
@KevinHwangAi ttyUSB0 is wrong for the firmware, it has to be /dev/ttyS0 for the first UART on the board.

**Comment by Viplav04** [comment]
It should be to /dev/ttyS0. /dev/ttyUSB0(or other) are the device names at Linux.

You can also enable debug features in debug options and check the problem but I don't know if it will give out info with micro-ROS. 
See video : https://youtu.be/A2kyHctm9YM

**Comment by KevinHwangAi** [comment]
@iluetkeb @Viplav04 Thank you.I will try again immediately.

**Comment by KevinHwangAi** [comment]
@iluetkeb @Viplav04  It works now!
Next, I'd like to try  the dashing version of mirco_ros.
Thank you again!!!

**Comment by Viplav04** [comment]
> @KevinHwangAi ttyUSB0 is wrong for the firmware, it has to be /dev/ttyS0 for the first UART on the board.

@iluetkeb maybe there is no issue with your default configuration as well, i was also setting the /dev/ttyS0 wrong that time and trying different configs and blinking light was only coming when executing publisher. Can this also cause a memory overflow?? The publisher messages not going to the agent?

**Comment by KevinHwangAi** [comment]
@iluetkeb I'm very much looking forward to a formal Milco Ros official tutorial.

**Comment by iluetkeb** [comment]
@Viplav04 No, we use fixed buffers, there won't be an overflow even when there is no agent.

At the moment I'm busy with the ROSCon demo, but a colleague is also already looking at the publisher demo. We'll see...

**Comment by Viplav04** [comment]
Hi @KevinHwangAi 
Did you try running publisher with UDP connection?

I get this error : nsh> publisher
modified publisher
UDP mode => ip: 10.0.0.1 - port: 8888

>>> [rcutils|error_handling.c:106] rcutils_set_error_state()
This error state is being overwritten:

  'failed to create node session on Micro ROS Agent., at /home/rosteam/uros_ws/firmware/mcu_ws/uros/rmw_microxrcedds/rmw_microxrcedds_c/src/rmw_node.c:235, at /home/rosteam/uros_ws/firmware/mcu_ws/ros2/rcl/rcl/src/rcl/node.c:402'

with this new error message:

  'rcl node's rmw handle is invalid, at /home/rosteam/uros_ws/firmware/mcu_ws/ros2/rcl/rcl/src/rcl/node.c:535'

rcutils_reset_error() should be called after error handling to avoid this.
<<<
[ERROR] [rcl]: Failed to fini publisher for node: 1
[rclc_create_node] error in rcl_node_init: rcl node's rmw handle is invalid, at /home/rosteam/uros_ws/firmware/mcu_ws/ros2/rcl/rcl/src/rcl/node.c:535
Issues creating node
----------------------------------------------------------------------------------------------------------------------------------
At the agent it says ok : 
~/uros_ws/install/micro_ros_agent/lib/micro_ros_agent$ ./micro_ros_agent udp 8888
UDP agent initialization... OK
Enter 'q' for exit

@BorjaOuterelo 
Have you found this error before and fixed it?
Maybe I am using old micro ROS Agent repos?

**Comment by KevinHwangAi** [comment]
@Viplav04 I using the serial connection. 
What version of the agent are you using?

**Comment by Viplav04** [comment]
@KevinHwangAi 
There were some installation issue for the agent before, so they created a seperate branch and then I was successful in installation, but I guess, now they have fixed that problem in the master. I am using the crystal version at both firmware and agent.

**Comment by KevinHwangAi** [comment]
@Viplav04 I am using the Dashing version，they said the agent may faster than the old version.

**Comment by KevinHwangAi** [comment]
@Viplav04 I want to try another MCU(stm32f4dis) to install the Mirco ros.
But it failed at configuration and compilation and  still trying to figure out ...

**Comment by iluetkeb** [comment]
@KevinHwangAi we have some C++-related settings in the Olimex-STM32-E407 BSP which are not common in the other BSPs. Please open a new issue, where you give some information on the board you want to use, and then I can help you further.
